lib/remote_node.go

```diff
@@ -0,0 +1,599 @@
+	case RemoteNodeStatus_Connected, RemoteNodeStatus_Validated:
+		rn.cmgr.CloseConnection(id)
+	}
+	rn.setStatusTerminated()
+}
+
+func (rn *RemoteNode) SendMessage(desoMsg DeSoMessage) error {
+	if rn.connectionStatus != RemoteNodeStatus_Connected && rn.connectionStatus != RemoteNodeStatus_Validated {
+		return fmt.Errorf("SendMessage: Remote node is not connected")
+	}
+
+	if err := rn.cmgr.SendMessage(desoMsg, rn.GetId().ToUint64()); err != nil {
+		return fmt.Errorf("SendMessage: Problem sending message to peer (id= %d): %v", rn.id, err)
+	}
+	return nil
+}
+
+// InitiateHandshake is a starting point for a peer handshake. If the peer is outbound, a version message is sent
+// to the peer. If the peer is inbound, the peer is expected to send a version message to us first.
+func (rn *RemoteNode) InitiateHandshake(nonce uint64) error {
+	if rn.connectionStatus != RemoteNodeStatus_Connected {
+		return fmt.Errorf("InitiateHandshake: Remote node is not connected")
+	}
+
+	if rn.GetPeer().IsOutbound() {
+		rn.versionTimeExpected = time.Now().Add(rn.params.VersionNegotiationTimeout)
+		return rn.sendVersionMessage(nonce)
+	}
+	return nil
+}
+
+// sendVersionMessage generates and sends a version message to a RemoteNode peer. The message will contain the nonce
+// that is passed in as an argument.
+func (rn *RemoteNode) sendVersionMessage(nonce uint64) error {
+	verMsg := rn.newVersionMessage(nonce)
+
+	// Record the nonce of this version message before we send it so we can
+	// detect self connections and so we can validate that the peer actually
+	// controls the IP she's supposedly communicating to us from.
+	vMeta := rn.getHandshakeMetadata()
+	vMeta.versionNonceSent = nonce
+
+	if err := rn.SendMessage(verMsg); err != nil {
+		return fmt.Errorf("sendVersionMessage: Problem sending version message to peer (id= %d): %v", rn.id, err)
+	}
+	return nil
+}
+
+// newVersionMessage returns a new version message that can be sent to a RemoteNode peer. The message will contain the
+// nonce that is passed in as an argument.
+func (rn *RemoteNode) newVersionMessage(nonce uint64) *MsgDeSoVersion {
+	ver := NewMessage(MsgTypeVersion).(*MsgDeSoVersion)
+
+	ver.Version = rn.params.ProtocolVersion.ToUint64()
+	// Set the services bitfield to indicate what services this node supports.
+	ver.Services = SFFullNodeDeprecated
+	if rn.hyperSync {
+		ver.Services |= SFHyperSync
+	}
+	if rn.archivalMode {
+		ver.Services |= SFArchivalNode
+	}
+	if rn.posValidator {
+		ver.Services |= SFPosValidator
+	}
+
+	// We use an int64 instead of a uint64 for convenience.
+	ver.TstampSecs = time.Now().Unix()
+
+	ver.Nonce = nonce
+	ver.UserAgent = rn.params.UserAgent
+
+	// When a node asks you for what height you have, you should reply with the height of the latest actual block you
+	// have. This makes it so that peers who have up-to-date headers but missing blocks won't be considered for initial
+	// block download.
+	ver.StartBlockHeight = uint32(rn.bc.BlockTip().Header.Height)
```

Comment:
Rather than reading the current height from the blockchain, see if you can pass the current height into the InitiateHandshake and HandleVersionMessage functions instead

Reply:
seconded!